src/misc/paper-write.ts

```diff
@@ -0,0 +1,321 @@
+    msg.sender.location,
+    'entity.villager.work_cartographer',
+    SoundCategory.PLAYERS,
+    1,
+    1,
+  );
+  msg.sender.inventory.removeItem(
+    new ItemStack(Material.INK_SAC, mainHand.amount),
+  );
+}
+
+/**
+ * Alternative interface for detecting writing paper hands.
+ */
+
+export function detectWritingPaper(msg: ChatMessage) {
+  if (playersWriting.has(msg.sender)) {
+    msg.discard = true;
+    handleMessage(msg);
+  }
+}
+
+GLOBAL_PIPELINE.addHandler('detectWritingPaper', 0, detectWritingPaper);
+
+/**
+ * Put paper inside envelope
+ */
+Envelope.event(
+  PlayerInteractEvent,
+  (event) => event.player.inventory.itemInOffHand,
+  async (event) => {
+    if (!isRightClick(event.action)) return;
+    if (event.hand !== EquipmentSlot.HAND) return;
+    if (event.item?.type !== Material.PAPER) return;
+    // Check if player has feather
+    const inventory = event.player.inventory as PlayerInventory;
+    const offHand = inventory.itemInOffHand;
+    const paperWritten = PaperWritten.get(event.item);
+    const paperSealed = PaperSealed.get(event.item);
+    if (paperWritten) {
+      if (event.item.amount > offHand.amount) {
+        event.player.sendActionBar(
+          `Ei tarpeeksi kirjekuoria (${event.item.amount})`,
+        );
+        return;
+      }
+      if (!event.item.itemMeta || !event.item.itemMeta.lore) return;
+      event.player.sendActionBar('§7Laitoit kirjeen kirjekuoreen');
+      const envelope = EnvelopeWithLetter.create({
+        letter: ChatColor.stripColor(event.item.itemMeta.lore[0]),
+        wax: [],
+      });
+      envelope.amount = event.item.amount;
+      inventory.itemInMainHand = envelope;
+      inventory.itemInOffHand.amount -= event.item.amount;
+    } else if (paperSealed) {
+      if (event.item.amount > offHand.amount) {
+        event.player.sendActionBar(
+          `Ei tarpeeksi kirjekuoria (${event.item.amount})`,
+        );
+        return;
+      }
+      if (
+        !event.item.itemMeta ||
+        !event.item.itemMeta.lore ||
+        event.item.itemMeta.lore.length != 3
+      )
+        return;
+      event.player.sendActionBar('§7Laitoit kirjeen kirjekuoreen');
+      const envelope = EnvelopeWithLetter.create({
+        letter: ChatColor.stripColor(event.item.itemMeta.lore[0]),
+        wax: [event.item.itemMeta.lore[1], event.item.itemMeta.lore[2]],
+      });
+      envelope.amount = event.item.amount;
+      inventory.itemInMainHand = envelope;
+      inventory.itemInOffHand.amount -= event.item.amount;
+    }
+  },
+);
+
+/**
+ * Open envelope
+ */
+
+EnvelopeWithLetter.event(
+  PlayerInteractEvent,
+  (event) => event.item,
+  async (event) => {
+    if (!isRightClick(event.action)) return;
+    if (event.hand !== EquipmentSlot.HAND) return;
+    if (event.item?.type !== Material.PAPER) return;
+    const inventory = event.player.inventory as PlayerInventory;
+    const offHand = inventory.itemInOffHand;
+    if (offHand.type !== Material.AIR) return;
```

Comment:
An error message for when offhand is full would be nice, it is not necessarily clear that you need two hands for opening an envelope.